apps/mccurtaind/mccurtaind.8
.Dd April 16, 2024
.Dt mccurtaind 8
.Os mcplex.net
.Sh NAME
.Nm mccurtaind
.Nd IP origin server daemon
.Sh SYNOPSIS
.Nm
.Op Fl c Ar config_file
.Op Fl p Ar pid_file
.Op Fl d
.Sh DESCRIPTION
.Nm
is a server daemon that provides routing origin information to clients
via UDP (for IP address queries) and secure TCP (for IP and other queries).
It requires a configuration file (see
.Xr mccurtaind.cfg 8 ), a mccurtain dtabase file (see
.Xr mkcurtaindb 8 ,
.Xr mccurtaindb 5 and
;Xr mccdb 1 ) and a RIPE \fIasn.txt\fR file.
.Pp
.Nm
uses libDwmCredence for authentication and encryption of TCP connections.
The UDP service in unauthenticated and ununecrypted, but significantly
faster (much lower latency) for the queries supported; it can be very
useful bound to only the loopback, pqrticularly on hosts performing
intrusion detection/prevention.
.Pp
The following command line options are available:
.Pp
.Bl -tag -width indent
.It Fl c Ar config_file
Specify the
.Xr mccurtaind 8
configurtion file.  The default path is \fI/usr/local/etc/mccurtaind.cfg\fR.
See
.Xr mccurtaind.cfg 5 for configuration documentation.
.It Fl p Ar pid_file
Specify the file in which
.Nm
should store its process ID (PID).  The default is
\fI/var/run/mccurtaind.pid\fR
.It Fl d
Do not daemonize.  Mostly useful for debugging.
.El

.Sh FILES
.Bl -tag -width indent
.It Pa /usr/local/etc/mccurtaind.cfg
The default configuration file.
.It Pa /var/run/mccurtaind.pid
The default PID file.
.El
.Sh SEE ALSO
.Xr mccurtain 1 ,
.Xr mccdb 1 ,
.Xr mccurtaindb 5 ,
.Xr mccurtaind.cfg 5 ,
.Xr mkcurtaindb 8 ,
.Xr credence 1
.Sh AUTHORS
Copyright \(co 2024
.An Daniel W. McRobb Aq Mt dwm@mcplex.net


ChangeLog
2026-09-17  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* tagged as mccurtain-0.0.38

	* apps/mccurtain/Makefile
	* apps/mccurtaind/Makefile
	* apps/mkcurtaindb/Makefile
	* classes/tests/Makefile
	Remove '-c' from compiler arguments when buulding dependencies.

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.37

	* etc/mccurtaind.cfg.example
	Correct a comment.

2026-09-17  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mccurtain-0.0.36

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/Makefile
	* apps/mccurtaind/mccurtaind.cfg.5
	Added manpage for the mccurtaind configuration file.

2026-09-16  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* tagged as mccurtain-0.0.35

	* apps/mkcurtaindb/Makefile
	Correct tarprep target path for mccurtaindb.5

	* tagged as mccurtain-0.0.34

2026-09-16  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mccurtaind/DwmMcCurtainUdpServer.cc
	* apps/mccurtaind/DwmMcCurtainUdpServer.hh
	Check client access in UDP server.

	* etc/mccurtaind.cfg.example
	Add udpAddresses configuration.

	* apps/mkcurtaindb/mccurtaindb.5
	Add manpage for db file format.

	* apps/mccurtaind/mccurtaind.8
	* apps/mccurtaind/Makefile
	Add manpage for mccurtaind(8)

	* apps/mkcurtaindb/mkcurtaindb.8
	* apps/mkcurtaindb/Makefile
	Add manpage for mkcurtaindb(8)

	* tagged as mccurtain-0.0.33

	* classes/include/DwmMcCurtainResponses.hh
	include DwmIpv6Prefix.hh

2026-09-15  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.32

	* apps/mccurtaind/DwmMcCurtainServer.cc
	Populate IPv6 prefixes in responses to AS or country queries.

	* classes/include/DwmMcCurtainResponses.hh
	Add std::vector<Ipv6Prefix> to ASPrefixesResponse so that responses
	to queries for an AS or country can contain IPv6 prefixes.

	* apps/mccurtain/mccurtain.cc
	Added -4 and -6 command line options to select IPv4 or IPv6 prefix
	output when querying for an AS or country.  The default selects only
	IPv4 prefixes.

	* apps/mccurtain/mccurtain.cc
	* apps/mccurtaind/DwmMcCurtainResponder.cc
	* apps/mccurtaind/DwmMcCurtainResponder.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* classes/include/DwmMcCurtainRequests.hh
	* classes/include/DwmMcCurtainResponses.hh
	Added support for IPv6 address queries via our secure channel.

2026-09-15  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* apps/mccdb/mccdb.cc
	Print country code when printing ASes within a country code.

2026-09-15  Daniel McRobb  <dwm@thrip.mcplex.net>

	* apps/mccdb/mccdb.1
	Added some examples.

	* apps/mccdb/mccdb.cc
	* apps/mccdb/mccdb.1
	Added '-4' and '-6' command line options to select only IPv4 or
	IPv6 prefixes.

2026-09-15  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Renamed mccdbcat to mccdb.

	* apps/mccdbcat/mccdbcat.cc
	Allow selecting prefixes which match a given address, a given
	country or given AS.  Also allow display of ASes with a
	given country code.

2026-09-13  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.31

	* apps/mccdbcat/mccdbcat.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/src/DwmMcCurtainDatabaseConfig.cc
	* etc/mccurtaind.cfg.example
	* scripts/updatecurtaindata.sh
	Be consistent about default db filename (now
	/usr/local/etc/mccas2ip.db).

	* tagged as mccurtain-0.0.30

	* apps/mccdbcat/Makefile
	* apps/mccdbcat/mccdbcat.1
	Added manpage for mccdbcat(1)

	* etc/mccurtaind.cfg.example
	Added to comment for dbFile.

	* apps/mccdbcat/mccdbcat.cc
	Use AS2IPDb::Load()

2026-09-13  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2IPDb.hh
	* classes/src/DwmMcCurtainAS2IPDb.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/Makefile
	* classes/tests/Makefile
	* classes/tests/TestAS2IPDb.cc
	Added AS2IPDb class with static members to load and save
	AS2Ipv4Net and AS2Ipv6Net via common interface that automatically
	uses bz2, gz or no compression based on filename extension.

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	Allow loading AS to IP data from bzip2 or gzip files based on
	file extension.

	* apps/mkcurtaindb/mkcurtaindb.cc
	If output file name extension is ".bz2", use BZ2 compression.
	If output file name extension is ".gz", use gzip compression.
	Else use no compression.

2026-09-12  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	Save() and Load() will now use bzip2 compression if path extension
	is ".bz2", gzip compression if path extension is ".gz", and no
	compression in all other cases.

2026-09-12  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainDnsServer.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/include/DwmMcCurtainIpv4Net2AS.hh
	* classes/include/DwmMcCurtainIpv6Net2AS.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/DwmMcCurtainIpv4Net2AS.cc
	* classes/src/DwmMcCurtainIpv6Net2AS.cc
	* classes/tests/.gitignore
	* classes/tests/Makefile
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	* classes/tests/TestCaidaV4Routeviews.cc
	* classes/tests/TestCaidaV6Routeviews.cc
	db file now only contains an AS2Ipv4Net and AS2Ipv6Net, since
	we can construct our Ipv4Net2AS and Ipv6Net2AS from them.  This
	reduces thr size of the uncompressed binaery bb file to a size
	less than the gzip'ed CAIDA routeviews files.

	I also added more I/O functionality to AS2Ipv4Net and AS2Ipv6Net

apps/mccurtaind/mccurtaind.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2024
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mccurtaind.cc
//!  \author Daniel W. McRobb
//!  \brief McCurtain server daemon
//---------------------------------------------------------------------------

#include <cstdio>
#include <cstdlib>

#include "DwmDaemonUtils.hh"
#include "DwmSignal.hh"
#include "DwmMclogLogger.hh"
#include "DwmMcCurtainVersion.hh"
#include "DwmMcCurtainServer.hh"

using namespace std;
using namespace Dwm;

static string            g_pidFile;
static atomic<bool>      g_gotSigTerm = false;
boost::asio::io_context  g_serverContext;

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void SavePID(const string & pidFile)
{
#ifndef O_EXLOCK
  #define O_EXLOCK 0
#endif
  g_pidFile = pidFile;
  pid_t  pid = getpid();
  string  &&pidstr = to_string(pid) + "\n";
  int    fd = open(g_pidFile.c_str(), O_WRONLY|O_CREAT|O_TRUNC|O_EXLOCK, 0644);
  if (fd >= 0) {
    write(fd, pidstr.c_str(), pidstr.size());
    close(fd);
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void RemovePID()
{
  if (! g_pidFile.empty()) {
    std::remove(g_pidFile.c_str());
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void SigTermHandler(int sigNum)
{
  if (SIGTERM == sigNum) {
    g_gotSigTerm = true;
    g_serverContext.stop();
    MCLOG(LOG_INFO, "SigTermHandler got SIGTERM");
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void BoostSignalHandler(const boost::system::error_code & ec,
                               int sigNum)
{
  if (SIGTERM == sigNum) {
    g_gotSigTerm = true;
    g_serverContext.stop();
    MCLOG(LOG_INFO, "BoostSignalHandler got SIGTERM");
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  cerr << "Usage: " << argv0 << " [-d] [-c configFile] [-p pidFile]\n"
       << "  Default configFile: /usr/local/etc/mccurtaind.cfg\n"
       << "  Default pidFile: /var/run/mccurtaind.pid\n"
       << "  -d: do not daemonize\n";
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  string  configFile("/usr/local/etc/mccurtaind.cfg");
  string  pidFile("/var/run/mccurtaind.pid");
  bool    daemonize = true;
  
  int  optChar;
  while ((optChar = getopt(argc, argv, "c:dp:")) != -1) {
    switch (optChar) {
      case 'c':
        configFile = optarg;
        break;
      case 'd':
        daemonize = false;
        break;
      case 'p':
        pidFile = optarg;
        break;
      default:
        Usage(argv[0]);
        return 1;
        break;
    }
  }

  McCurtain::Config  config;
  if (config.Parse(configFile)) {
    if (daemonize) {
      DaemonUtils::Daemonize();
    }
    Dwm::Mclog::logger.Open(config.SyslogFacility());
    Dwm::Mclog::logger.MinimumSeverity(config.SyslogLevel());
    Dwm::Mclog::logger.LogLocations(config.SyslogLocations());

    Signal sigHup(SIGTERM);
    sigHup.PushHandler(SigTermHandler);
    SavePID(pidFile);
    boost::asio::signal_set boostSignals(g_serverContext, SIGTERM);
    boostSignals.async_wait(BoostSignalHandler);
    McCurtain::Server  server(g_serverContext.get_executor(), config);
    g_serverContext.run();
    for (;;) {
      if (g_gotSigTerm) {
        server.Stop();
        RemovePID();
        break;
      }
      else {
        this_thread::sleep_for(chrono::milliseconds(500));
      }
    }
  }
  else {
    cerr << "failed to parse configuration in '" << configFile << "'\n";
    return 1;
  }
  
  return 0;
}
